### PR TITLE
Support raw devfile urls without yaml extension

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
On handling raw devfile urls, request content by the url, and check if the content is a devfile. If yes treat the url as a raw devfile url.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse-che/che/issues/22937

### How to test this PR?

1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-683`
2. Start a workspace from a raw devfile url without `yaml` extension, e.g. `https://raw.githubusercontent.com/vinokurig/test/master/parent`

See: workspace starts successfully.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/RawDevfileUrlFactoryParameterResolver.java [hunk 1]
```diff
@@ -64,7 +69,17 @@ public RawDevfileUrlFactoryParameterResolver(
   @Override
   public boolean accept(Map<String, String> factoryParameters) {
     String url = factoryParameters.get(URL_PARAMETER_NAME);
-    return !isNullOrEmpty(url) && PATTERN.matcher(url).matches();
+    return !isNullOrEmpty(url) && PATTERN.matcher(url).matches() || containsDevfile(url);
```
Comment (tolusha) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/RawDevfileUrlFactoryParameterResolver.java:72]:
```suggestion
    return !isNullOrEmpty(url) && (PATTERN.matcher(url).matches() || containsDevfile(url));
```

**Review by tolusha — APPROVED** [review]

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/RawDevfileUrlFactoryParameterResolver.java [hunk 2]
```diff
@@ -64,7 +69,17 @@ public RawDevfileUrlFactoryParameterResolver(
...
+  }
+
+  private boolean containsDevfile(String requestURL) {
+    try {
+      String fetch = urlFetcher.fetch(requestURL);
+      devfileParser.parseYaml(fetch);
+      return true;
+    } catch (IOException | DevfileFormatException e) {
+      return !e.getMessage().startsWith("Cannot construct instance of");
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/RawDevfileUrlFactoryParameterResolver.java]: reworked to parsing raw yaml

**Review by ibuziuk — APPROVED** [review]

**Comment by artaleks9** [comment]
The `gitlab-with-oauth-setup-flow` is failed.
Maybe, it's related to UI locator change
I need to investigate it.

**Comment by artaleks9** [comment]
@vinokurig the `gitlab-with-oauth-setup-flow` is fail due to the known issue -> https://github.com/eclipse-che/che/issues/22953
The `Views and More Actions` menu is not available (see the screenshot)

---

@RomanNikitenko said this problem was resolved in the `next` 4 days ago, but, it seems this `che-sever` image is not included the fix. Do we have an opportunity update the PR image, to rerun the `gitlab-with-oauth-setup-flow` test and verify it?

----

Anyway I've checked the functionality related to the PR and OAuth - works properly.

che-pr-683

**Comment by vinokurig** [comment]
@artaleks9 
> Do we have an opportunity update the PR image, to rerun the gitlab-with-oauth-setup-flow test and verify it?

The PR image is already up to date with the `main` branch.

**Comment by vinokurig** [comment]
@artaleks9 If the failure is a known issue, can we go with the current changes?

**Comment by RomanNikitenko** [comment]
I see on the screenshot that editor version is pretty old - it was built 2 weeks ago.
I think editor image should be updated for testing.

**Comment by artaleks9** [comment]
| @artaleks9 If the failure is a known issue, can we go with the current changes?
I think, yes

**Review by artaleks9 — APPROVED** [review]

**Pull request merged**